Allow `./mach build` to enable Servo's headless mode

nerith opened the pull request:
As mentioned in #7512 and #7637, the environment variable SERVO_HEADLESS
enables building the OSMesa version of Servo when it is set to 1.

Instead, this should be a command line option to mach's build command.

Fixes #7637.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7734)

<!-- Reviewable:end -->

metajack commented:
That sure was simpler than I expected :) A few comments to address.
<!-- Reviewable comment -K-KfaXgxgYdLgGdoTh0:-1398911605 -->

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[python/servo/build_commands.py, line 135 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfTtAzP-9eb6PKGxr)** ([raw file](https://github.com/servo/servo/blob/9b87495614d8d7c4ab2622fc27e12a37d4b7a8c6/python/servo/build_commands.py#L135)):</sup>
This is linux only correct? Should we also warn if you try to use this on another platform?
<!-- Reviewable comment -K-KfTtAzP-9eb6PKGxs:2014224601 -->

---

<sup>**[python/servo/build_commands.py, line 207 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfQelTTj8ZUnTDpL9)** ([raw file](https://github.com/servo/servo/blob/d811f2d1baac6dcb501f3588eb509086bd1b2a66/python/servo/build_commands.py#L207)):</sup>
If we're not calling this function we can probably delete it now.
<!-- Reviewable comment -K-KfQelTTj8ZUnTDpLA:-2132235884 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7734)
<!-- Sent from Reviewable.io -->

nerith commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[python/servo/build_commands.py, line 135 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfTtAzP-9eb6PKGxr)** ([raw file](https://github.com/servo/servo/blob/9b87495614d8d7c4ab2622fc27e12a37d4b7a8c6/python/servo/build_commands.py#L135)):</sup>
I believe this is linux only. Would a warning such as "Headless mode (OSMesa) is not supported on your platform." be sufficient if the platform isn't linux?
<!-- Reviewable comment -K-TRG_YYDlBwMBfDqhW:214391532 -->

---

<sup>**[python/servo/build_commands.py, line 207 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfQelTTj8ZUnTDpL9)** ([raw file](https://github.com/servo/servo/blob/d811f2d1baac6dcb501f3588eb509086bd1b2a66/python/servo/build_commands.py#L207)):</sup>
We're still calling is_headless_build() for the build-tests command.
<!-- Reviewable comment -K-TQhK4mxvzKobHtE2p:-468780746 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7734)
<!-- Sent from Reviewable.io -->

metajack commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[python/servo/build_commands.py, line 135 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfTtAzP-9eb6PKGxr)** ([raw file](https://github.com/servo/servo/blob/9b87495614d8d7c4ab2622fc27e12a37d4b7a8c6/python/servo/build_commands.py#L135)):</sup>
That sounds great.
<!-- Reviewable comment -K-TgdSfmsvl1YROjxJa:360540361 -->

---

<sup>**[python/servo/build_commands.py, line 207 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7734#-K-KfQelTTj8ZUnTDpL9)** ([raw file](https://github.com/servo/servo/blob/d811f2d1baac6dcb501f3588eb509086bd1b2a66/python/servo/build_commands.py#L207)):</sup>
is_headless_build is still using the env var, which is wrong. So if it's still called elsewhere we should fix it to do the right thing.

This probably means that --headless will have to stuff something into the context so that it can be pulled out. Or perhaps it's enough to have build-tests take --headless.
<!-- Reviewable comment -K-TgfEaXoJIwN4-3jub:-488608227 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7734)
<!-- Sent from Reviewable.io -->

metajack commented:
Looks good; just had one nit. When that's fixed r=me.
<!-- Reviewable comment -K0XEWODp9G6NVsUxCmk:1158053181 -->

---

Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[python/servo/build_commands.py, line 353 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7734#-K0XEMZoHoroVYq0xoaB)** ([raw file](https://github.com/servo/servo/blob/ef5cbc832be21ec3d6da81639f30fa4e708c4be7/python/servo/build_commands.py#L353)):</sup>
nit: if headless and headless_supported() would work here.
<!-- Reviewable comment -K0XEMZoHoroVYq0xoaC:-1608643830 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7734)
<!-- Sent from Reviewable.io -->

Manishearth commented:
@nerith Updates on this? Need any help?

nerith commented:
@Manishearth This is done.

nerith commented:
r= @metajack

Manishearth commented:
@bors-servo r=jack
<!-- Reviewable comment -K0Xm4rPaipGLc5ElIa9:-96637707 -->

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7734)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 8012974 has been approved by `jack`

<!-- @bors-servo r=jack 80129748ec0a57ca95a9eaa08fc6fee09fbd822e -->

bors-servo commented:
:hourglass: Testing commit 8012974 with merge be6e7dd...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/524)

Manishearth commented:
@bors-servo retry

(submodule clone error)

bors-servo commented:
:zap: Previous build results for [gonk](http://build.servo.org/builders/gonk/builds/2572) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/3412), [linux-dev](http://build.servo.org/builders/linux-dev/builds/490), [linux-rel](http://build.servo.org/builders/linux-rel/builds/507), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/524), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/506), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/541)...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/490)

Manishearth commented on python/servo/build_commands.py:
```diff
@@ -25,8 +24,14 @@
 from servo.command_base import CommandBase, cd
 
 
-def is_headless_build():
```
This function shouldn't be removed, we still need `$SERVO_HEADLESS` to work.

We should build headless if either `--headless` is passed or if `$SERVO_HEADLESS` is defined.

metajack replied:
Why do we need $SERVO_HEADLESS to work? I want to kill that with fire!

Manishearth replied:
We can't land this PR without it, and we can't change the CI without this PR. I'm saying we do this step by step :)

But closing the tree, updating CI, and landing this would work too.

metajack replied:
Ah yes. I'll do that in a bit I guess just to save the extra few roundtrips.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8176) made this pull request unmergeable. Please resolve the merge conflicts.

nerith committed:
Allow `./mach build` to enable Servo's headless mode

As mentioned in #7512 and #7637, the environment variable SERVO_HEADLESS
enables building the OSMesa version of Servo when it is set to 1.

Instead, this should be a command line option to mach's build command.

Fixes #7637.

nerith committed:
Allow `./mach build-tests` to enable Servo's headless mode

This allows mach's build-tests command to use the option --headless
instead of checking for the value of SERVO_HEADLESS.

jdm commented:
@bors-servo: r=metajack

bors-servo commented:
:pushpin: Commit 2342b1b has been approved by `metajack`

<!-- @bors-servo r=metajack 2342b1bc6f068cc70302c26a7c06d72a31ca0f91 -->

bors-servo commented:
:hourglass: Testing commit 2342b1b with merge 3c0cd5e...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/753), [gonk](http://build.servo.org/builders/gonk/builds/769), [linux-dev](http://build.servo.org/builders/linux-dev/builds/752), [linux-rel](http://build.servo.org/builders/linux-rel/builds/833), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/775), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/761), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/787)
